### DISCUSSION TITLE
Run "remove unused imports" on format

### DISCUSSION THREAD

**Discussion by andrefgneves** [zed-industries/zed#13403] [Feature Requests]
I have the following setting:

```json
"code_actions_on_format": {
  "source.fixAll.eslint": true
}
```

but I also want to run the "remove unused imports" action. Is there a way to do that?

**Comment by andrefgneves** [comment]
For anyone interested, I found the setting:

```
"code_actions_on_format": {
  "source.fixAll.eslint": true,
  "source.removeUnusedImports": true
}
```

**Reply by samiursakib** [reply]
In my case, this code snippet wasn't able to remove the unused imports from the buffer unless `source.organizeImports` was used. Isn't it a weird behavior of zed in my machine? @andrefgneves

**Reply by Melvynx** [reply]
this is not working for me, the eslint is apply but the removeUnusedImports is not

**Comment by glassworks-projects** [comment]
this is not working for me.

**Reply by sebkolind** [reply]
@glassworks-projects did you enable `source.organizeImports` as https://github.com/zed-industries/zed/discussions/13403#discussioncomment-10926979 wrote? That did the trick for me 🎉

**Reply by glassworks-projects** [reply]
yes. still does not work.

**Reply by sebkolind** [reply]
> yes. still does not work.

What TS version are you on? I think this is a feature that was added in 5.1 or something like that.

**Reply by glassworks-projects** [reply]
```
"typescript": "^5.1.6",
```
i think actual version is `5.7.3`

**Reply by sebkolind** [reply]
> ```
> 
> "typescript": "^5.1.6",
> 
> ```
> 
> i think actual version is `5.7.3`

Just checked, and the required TS version is 4.9. However, it seems odd that it wouldn't remove the unused imports for you. It is working for me but only with the other option too, which is also not very nice.

I'd like to be able to remove unused imports without sorting.

**Comment by NunciosChums** [comment]
This is how I do it.  I found it at https://markobajlovic.com/journal/zed-settings-in-2025

```json 
"code_actions_on_format": {
  "source.organizeImports": true,
  "source.fixAll.ruff": true
}
```

**Comment by ziimakc** [comment]
If `source.organizeImports` makes your imports jump or it show changes when there is none do the following:

in `.zed/settings.json`
```
	"code_actions_on_format": {
		"source.fixAll.biome": true,
		"source.organizeImports.biome": true
	}
}
```

in biome.json:
```
{
	"$schema": "./node_modules/@biomejs/biome/configuration_schema.json",
	"files": {
		"ignoreUnknown": true
	},
	"linter": {
		"includes": [
			"**",
			"!**/generated/**",
			"!**/scripts/**",
			"!apps/frontend/src/tailwind.css"
		],
			"correctness": {
				"noUnusedImports": { "fix": "safe", "level": "error" } // this makes `source.fixAll.biome` remove unused imports
			},
		},
	},
	"formatter": {
		"enabled": true,
		"includes": ["**", "!**/generated/**"]
	},
	"assist": {
		"actions": { "source": { "organizeImports": "on" } }, // this will organize imports
		"includes": ["**", "!**/generated/**"]
	}
}
```

**Reply by Gin-Quin** [reply]
Applying { "fix": "safe" } did the trick 💪

**Reply by devxdemon** [reply]
This config worked for me. Thank you very much. This fixed one of the two features that is making me from using zed editor

**Reply by Amir-R1384** [reply]
What's the other one?
> This config worked for me. Thank you very much. This fixed one of the two features that is making me from using zed editor

**Reply by devxdemon** [reply]
Git panel with staged and unstaged diff
